src/main_em.cpp

```diff
@@ -236,53 +274,135 @@ void action_callback(ratslam_ros::TopologicalActionConstPtr action, ratslam::Exp
 #endif
 }
 
-void set_goal_pose_callback(geometry_msgs::PoseStampedConstPtr pose, ratslam::ExperienceMap * em)
+void set_goal_pose_callback(geometry_msgs::PoseStampedConstPtr pose)
 {
   em->add_goal(pose->pose.position.x, pose->pose.position.y);
+}
+
+void tf_update_timer_callback(const ros::TimerEvent& event)
+{
+  tf::StampedTransform odom_to_base_;
+  try
+  {
+    tf_listener->lookupTransform(odom_frame, base_frame, ros::Time(0), odom_to_base_);
+  }
+  catch (tf::TransformException& e)
+  {
+    ROS_WARN("Failed to compute odom pose. Exception: %s", e.what());
+    return;
+  }
+  tf::Transform odom_to_base = odom_to_base_;
+
+  tf::Transform base_to_map =
+      tf::Transform(tf::createQuaternionFromRPY(0, 0, mpose[2]), tf::Vector3(mpose[0], mpose[1], 0.0)).inverse();
+
+  map_to_odom = (odom_to_base * base_to_map).inverse();
 
+  ros::Time tf_expiration = ros::Time::now() + ros::Duration(0.5);
+  tf_broadcaster->sendTransform(tf::StampedTransform(map_to_odom, tf_expiration, map_frame, odom_frame));
 }
 
-int main(int argc, char * argv[])
+void save_map_timer_callback(const ros::TimerEvent& event)
 {
-  ROS_INFO_STREAM(argv[0] << " - openRatSLAM Copyright (C) 2012 David Ball and Scott Heath");
-  ROS_INFO_STREAM("RatSLAM algorithm by Michael Milford and Gordon Wyeth");
-  ROS_INFO_STREAM("Distributed under the GNU GPL v3, see the included license file.");
+  // create and open a character archive for output
+  std::ofstream ofs(map_file_path.c_str());
+
+  // save map to archive file
+  {
+    boost::archive::binary_oarchive binary_archive(ofs);
+    // write class instance to archive
+    binary_archive << *em;
+    // archive and stream closed when destructors are called
+  }
+}
+
+bool load_map(const std::string& file_path)
+{
+  try
+  {
+    // create and open an archive for input
+    std::ifstream ifs(file_path.c_str());
+    boost::archive::binary_iarchive binary_archive(ifs);
+    // read class state from archive
+    binary_archive >> *em;
+    // archive and stream closed when destructors are called
+    return true;
+  }
+  catch (...)
+  {
+    return false;
+  }
+}
+
+int main(int argc, char* argv[])
+{
+  // print info without ROS because it is not initialized yet
+  std::cout << argv[0] << " - openRatSLAM Copyright (C) 2012 David Ball and Scott Heath";
+  std::cout << "RatSLAM algorithm by Michael Milford and Gordon Wyeth";
+  std::cout << "Distributed under the GNU GPL v3, see the included license file.";
 
   if (argc < 2)
   {
-    ROS_FATAL_STREAM("USAGE: " << argv[0] << " <config_file>");
+    std::cout << "USAGE: " << argv[0] << " <config_file>";
```

Comment:
Checked it. I changed it because the ROS node is not initialized before. These macros make use of rosout publisher which is initialized with node or first node handle.